[terra-alert] Update alert roles for a11y (#3789)
cerner/terra-core: packages/terra-core-docs/src/terra-dev-site/doc/alert/AccessibilityGuide.2.doc.mdx

import { Notice } from "@cerner/terra-docs";
import { Badge } from 'terra-image/package.json?dev-site-package';
import Whitespace from '../../common/layout-helpers/Whitespace';

import Hyperlink from 'terra-hyperlink';

<Badge />

# Accessibility Guide for Terra Alert

<Notice variant="important" ariaLevel="2">

Using notifications improperly may prevent some of your users from understanding your UI.

</Notice>

<Whitespace />

## Why is this important?

Terra's Alert component is a notification banner with varying criticalities that communicates cautions, errors, information, and other general information to the user in context of their workflow. As such, special consideration must be made to inform screen reader users of their presence on the page, just as sighted users' attention would be drawn to them in the visual UI.

## Accessibility Considerations

### Code Considerations

#### Consistency

The use of each variant of this component should be consistent in order to provide the user the best experience.

#### Notification Criticality

The user only expects to be interrupted for notifications (alerts) that are of high criticality.

> * This behavior is referred to as "assertive".

Author: That's a good point! I feel like I should already be in the a11y mindset already but got that updated plus some really small changes to wording here https://github.com/cerner/terra-core/pull/3789/commits/c72d651509af567e23cc0bc0a0de59ad3f70504a#diff-2c4657d7282dff34f1995cecbdc9b487c519f018091c12370d911c9c60efd472

import { Notice } from "@cerner/terra-docs";
import { Badge } from 'terra-image/package.json?dev-site-package';
import Whitespace from '../../common/layout-helpers/Whitespace';

import Hyperlink from 'terra-hyperlink';

<Badge />

# Accessibility Guide for Terra Alert

<Notice variant="important" ariaLevel="2">

Using alerts improperly may prevent some of your users from understanding your UI.

</Notice>

<Whitespace />

## Why is this important?

> Alert is a form of notifications with varying criticalities that should inform and assist the user with their task.

Author: Updated here https://github.com/cerner/terra-core/pull/3789/commits/a7ad60dd792651447e2f6adbbe6d2b994da18b67#diff-2c4657d7282dff34f1995cecbdc9b487c519f018091c12370d911c9c60efd472R21

> > "An alert is an element that displays a brief, important message in a way that attracts the user's attention without interrupting the user's task."

Author: Removed this section here https://github.com/cerner/terra-core/pull/3789/commits/a7ad60dd792651447e2f6adbbe6d2b994da18b67#diff-2c4657d7282dff34f1995cecbdc9b487c519f018091c12370d911c9c60efd472L23

> _&nbsp;&nbsp;&mdash; excerpt from <Hyperlink href="https://www.w3.org/WAI/ARIA/apg/patterns/alert/" variant="external">ARIA Authoring Practices Guide (APG): Alert Pattern</Hyperlink>_[<sup>[1]</sup>](/components/cerner-terra-core-docs/alert/accessibility-guide#linked-references)

## Accessibility Considerations

### Code Considerations

#### Consistency

> The use of an alert notification should be consistent in order to provide the user the best experience.

Author: Updated here https://github.com/cerner/terra-core/pull/3789/commits/a7ad60dd792651447e2f6adbbe6d2b994da18b67#diff-2c4657d7282dff34f1995cecbdc9b487c519f018091c12370d911c9c60efd472R29

#### Alert Criticality

> The user only expects to be interrupted for notifications (alerts) that are of high criticality.

Author: Added here https://github.com/cerner/terra-core/pull/3789/commits/a7ad60dd792651447e2f6adbbe6d2b994da18b67#diff-2c4657d7282dff34f1995cecbdc9b487c519f018091c12370d911c9c60efd472R33

> Less critical notifications should wait until the current task is complete before presenting to the user.

Author: Added here https://github.com/cerner/terra-core/pull/3789/commits/a7ad60dd792651447e2f6adbbe6d2b994da18b67#diff-2c4657d7282dff34f1995cecbdc9b487c519f018091c12370d911c9c60efd472R34

By default, Terra Alert will set the role to `"alert"` for the `alert` type and `"status"` for all other types.
Terra Alert has an optional role prop but should be used carefully when the default does not meet accessibility or usability expectations.

| Role | Purpose |
| --- | --- |

> | Alert | See <Hyperlink href="https://w3c.github.io/aria/#alert" variant="external">Alert</Hyperlink>[<sup>[2]</sup>](/components/cerner-terra-core-docs/alert/accessibility-guide#linked-references) - "A type of live region with important, and usually time-sensitive, information." |

Reviewer: ```suggestion
| Alert | See <Hyperlink href="https://w3c.github.io/aria/#alert" variant="external"> WAI-ARIA Alert Role</Hyperlink>[<sup>[2]</sup>](/components/cerner-terra-core-docs/alert/accessibility-guide#linked-references) - "A type of live region with important, and usually time-sensitive, information." |
```

Author: Added here https://github.com/cerner/terra-core/pull/3789/commits/a7ad60dd792651447e2f6adbbe6d2b994da18b67#diff-2c4657d7282dff34f1995cecbdc9b487c519f018091c12370d911c9c60efd472R41

Author: Misread the suggestion, got that updated here https://github.com/cerner/terra-core/commit/6e3ba6b19f3867431af24f034ac3cfba3779bd14#diff-2c4657d7282dff34f1995cecbdc9b487c519f018091c12370d911c9c60efd472R41

> | Status | See <Hyperlink href="https://w3c.github.io/aria/#status" variant="external">Status</Hyperlink>[<sup>[2]</sup>](/components/cerner-terra-core-docs/alert/accessibility-guide#linked-references) - "A type of live region whose content is advisory information for the user but is not important enough to justify an alert, often but not necessarily presented as a status bar." |

Reviewer: ```suggestion
| Status | See <Hyperlink href="https://w3c.github.io/aria/#status" variant="external">WAI-ARIA Status Role</Hyperlink>[<sup>[2]</sup>](/components/cerner-terra-core-docs/alert/accessibility-guide#linked-references) - "A type of live region whose content is advisory information for the user but is not important enough to justify an alert, often but not necessarily presented as a status bar." |
```

Author: Added here https://github.com/cerner/terra-core/pull/3789/commits/a7ad60dd792651447e2f6adbbe6d2b994da18b67#diff-2c4657d7282dff34f1995cecbdc9b487c519f018091c12370d911c9c60efd472R42

Author: Misread the suggestion, got that updated here https://github.com/cerner/terra-core/pull/3789/commits/6e3ba6b19f3867431af24f034ac3cfba3779bd14#diff-2c4657d7282dff34f1995cecbdc9b487c519f018091c12370d911c9c60efd472R41

#### Assistive technologies

In combination with the Alert Criticality, screen reader users should understand the importance of the alert when presented.

The `"alert"` role will interrupt the user and announce the high critical notification immediately.
The `"status"` role will not interrupt the user and present politely.

<Notice variant="important" ariaLevel="5">

###### Accessibility Guidance: Live regions

> For screen readers to pick up and announce a newly rendered alert, the alert must be within an live region.

Reviewer: ```suggestion
For screen readers to pick up and announce a newly rendered notification, the notification must be within an aria-live region.
```

Author: Updated here https://github.com/cerner/terra-core/pull/3789/commits/a7ad60dd792651447e2f6adbbe6d2b994da18b67#diff-2c4657d7282dff34f1995cecbdc9b487c519f018091c12370d911c9c60efd472R55
